nemo_text_processing/inverse_text_normalization/vi/taggers/date.py

```diff
@@ -19,143 +19,143 @@
 
-    graph = pynini.union(
-        ties_graph + optional_ten + delete_space + (graph_digit | graph_one | graph_four | graph_five),
-        ties_graph + delete_space + graph_ten + pynutil.insert("0"),
-    )
-    return graph
+    def __init__(self, cardinal: GraphFst):
+        super().__init__(name="date", kind="classify")
 
+        cardinal_graph = cardinal.graph_no_exception
+        YEAR_WEIGHT = 0.001
 
-def _get_year_graph():
-    """
-    Transducer for year, e.g. hai không hai mươi -> 2020
-    """
+        graph_teen = pynini.string_file(get_abs_path("data/numbers/teen.tsv")).optimize()
+        graph_digit = pynini.string_file(get_abs_path("data/numbers/digit.tsv")).optimize()
+        graph_zero = pynini.string_file(get_abs_path("data/numbers/zero.tsv")).optimize()
+        ties_graph = pynini.string_file(get_abs_path("data/numbers/ties.tsv")).optimize()
+
+        # Special digit mappings for Vietnamese
+        graph_one = pynini.cross("mốt", "1")
+        graph_four = pynini.cross("tư", "4")
+        graph_five = pynini.cross("lăm", "5")
+        graph_ten = pynini.cross("mươi", "")
+        optional_ten = pynini.closure(delete_space + graph_ten, 0, 1)
+        # Ties graph for 20-99 (e.g., "hai ba" -> "23")
+        graph_ties = pynini.union(
+            ties_graph + optional_ten + delete_space + pynini.union(graph_digit, graph_one, graph_four, graph_five),
+            ties_graph + delete_space + graph_ten + pynutil.insert("0"),
+        )
 
-    def _get_digits_graph():
+        # Zero prefix patterns (e.g., "linh năm" -> "05")
         zero = pynini.cross((pynini.union("linh", "lẻ")), "0")
-        four = pynini.cross("tư", "4")
-        graph = pynini.union(
-            zero + delete_space + (graph_digit | four),
+        graph_digits = pynini.union(
+            zero + delete_space + pynini.union(graph_digit, graph_four),
             graph_zero + delete_space + graph_digit,
-        )
-        graph.optimize()
-        return graph
+        ).optimize()
 
-    def _get_hundreds_graph(graph_ties, graph_digits):
-        graph = (
+        # Year component builders
+        # Hundreds pattern (e.g., "hai trăm mười hai" -> "212")
+        graph_hundreds = (
             graph_digit
             + delete_space
             + pynutil.delete("trăm")
             + delete_space
-            + (graph_teen | graph_ties | graph_digits)
+            + pynini.union(graph_teen, graph_ties, graph_digits)
         )
-        return graph
 
-    def _get_thousands_graph(graph_ties, graph_digits):
-        graph_hundred_component = (
-            (graph_digit | graph_zero) + delete_space + pynutil.delete("trăm")
-        ) | pynutil.insert("0")
-        graph = (
+        # Thousands pattern (e.g., "hai nghìn không ba" -> "2003")
+        graph_hundred_component = pynini.union(
+            pynini.union(graph_digit, graph_zero) + delete_space + pynutil.delete("trăm"), pynutil.insert("0")
```

Comment:
without weighting you're going to get non-determinate behavior where a 0 is just inserted here.